pkg/admission/customresourcevalidation/network/validate_network.go

```diff
@@ -0,0 +1,130 @@
+package network
+
+import (
+	"fmt"
+	"io"
+	"net"
+
+	"k8s.io/apimachinery/pkg/api/validation"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/apiserver/pkg/admission"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/origin/pkg/admission/customresourcevalidation"
+)
+
+const PluginName = "config.openshift.io/ValidateNetwork"
+
+// Register registers a plugin
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
+		return customresourcevalidation.NewValidator(
+			map[schema.GroupResource]bool{
+				configv1.Resource("networks"): true,
+			},
+			map[schema.GroupVersionKind]customresourcevalidation.ObjectValidator{
+				configv1.GroupVersion.WithKind("Network"): networkV1{},
+			})
+	})
+}
+
+func toNetworkV1(uncastObj runtime.Object) (*configv1.Network, field.ErrorList) {
+	if uncastObj == nil {
+		return nil, nil
+	}
+
+	allErrs := field.ErrorList{}
+
+	obj, ok := uncastObj.(*configv1.Network)
+	if !ok {
+		return nil, append(allErrs,
+			field.NotSupported(field.NewPath("kind"), fmt.Sprintf("%T", uncastObj), []string{"Network"}),
+			field.NotSupported(field.NewPath("apiVersion"), fmt.Sprintf("%T", uncastObj), []string{"config.openshift.io/v1"}))
+	}
+
+	return obj, nil
+}
+
+type networkV1 struct {
+}
+
+// This only validates the syntax; the operator will worry about the semantics
+func validateNetworkSpec(spec configv1.NetworkSpec) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if len(spec.ClusterNetwork) == 0 {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec").Child("clusterNetwork"), nil, "must have at least 1 entry"))
+	}
+	for i, cnet := range spec.ClusterNetwork {
+		_, cidr, err := net.ParseCIDR(cnet.CIDR)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("spec").Child("clusterNetwork").Index(i).Child("cidr"), cnet.CIDR, err.Error()))
+		}
+		_, bits := cidr.Mask.Size()
+		if cnet.HostPrefix > uint32(bits) {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("spec").Child("clusterNetwork").Index(i).Child("hostPrefix"), cnet.HostPrefix, "too large for address type"))
+		}
+	}
+
+	if len(spec.ServiceNetwork) != 1 {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec").Child("serviceNetwork"), nil, "must have exactly 1 entry"))
+	}
+	for i, snet := range spec.ServiceNetwork {
+		_, _, err := net.ParseCIDR(snet)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("spec").Child("serviceNetwork").Index(i), snet, err.Error()))
+		}
+	}
+
+	if spec.NetworkType == "" {
```

Comment:
This is an enum right?  Validate as such.  Also we really should update to match the naming and structure of the Authentication type field.

Reply:
It is an enum, but we will be adding new elements to it over time, hopefully without needing other changes to this type. (This is just for "basic" config; "complicated" config happens in the network operator's config type.)

We (@openshift/sig-networking) probably need to think about the details of how this is going to work some more... Maybe it really only makes sense to allow `"OpenShiftSDN"`, `"OVNKubernetes"`, and `"Other"` in the config.openshift.io object, since if you're using a non-openshift-default networking type, it's pretty much guaranteed that you'll need to specify additional configuration options in the network operator config anyway...

Reply:
Consider `Authentication` as an example (and note that the name of the json field is `type` and not `authenticationType`:

```go
type AuthenticationSpec struct {
	// type identifies the cluster managed, user facing authentication mode in use.
	// Specifically, it manages the component that responds to login attempts.
	// The default is IntegratedOAuth.
	Type AuthenticationType `json:"type"`
...
```

```go
type AuthenticationType string

const (
	// None means that no cluster managed authentication system is in place.
	// Note that user login will only work if a manually configured system is in place and
	// referenced in authentication spec via oauthMetadata and webhookTokenAuthenticators.
	AuthenticationTypeNone AuthenticationType = "None"

	// IntegratedOAuth refers to the cluster managed OAuth server.
	// It is configured via the top level OAuth config.
	AuthenticationTypeIntegratedOAuth AuthenticationType = "IntegratedOAuth"

	// TODO if we add support for an in-cluster operator managed Keycloak instance
	// AuthenticationTypeKeycloak AuthenticationType = "Keycloak"
)
```

```go
// OAuth holds cluster-wide information about OAuth.  The canonical name is `cluster`.
// It is used to configure the integrated OAuth server.
// This configuration is only honored when the top level Authentication config has type set to IntegratedOAuth.
type OAuth struct {
... 
}
```

The only configuration we handle through the operator is the standard `OperatorSpec` and `GenericOperatorConfig`.